main_model.py

```diff
@@ -0,0 +1,101 @@
+        
+    else:
+        file_type = "unkown"
+        
+    return file_type
+
+from langchain.document_loaders.image import UnstructuredImageLoader
+from langchain.document_loaders import UnstructuredFileLoader
+
+def extract_file_content(file_path):
+    
+    file_type = detect_document_type(file_path)
+    
+    if(file_type == "pdf"):
+        loader = UnstructuredFileLoader(file_path)
+        
+    elif(file_type == "image"):
+        loader = UnstructuredImageLoader(file_path)
+        
+    documents = loader.load()
+    documents_content = '\n'.join(doc.page_content for doc in documents)
+    
+    return documents_content
+
+
+research_paper_path = "./data/transformer_paper.pdf"
+article_information_path = "./data/zoumana_article_information.png"
+
+print(f"Research Paper Type: {detect_document_type(research_paper_path)}")
+print(f"Article Information Document Type: {detect_document_type(article_information_path)}")
+
+research_paper_content = extract_file_content(research_paper_path)
+article_information_content = extract_file_content(article_information_path)
+
+nb_characters = 400
+
+print(f"First {nb_characters} Characters of the Paper: \n{research_paper_content[:nb_characters]} ...")
+print("---"*5)
+print(f"First {nb_characters} Characters of Article Information Document :\n {research_paper_content[:nb_characters]} ...")
+
+text_splitter = CharacterTextSplitter(        
+    separator = "\n\n",
+    chunk_size = 1000,
+    chunk_overlap  = 200,
+    length_function = len,
+)
+
+research_paper_chunks = text_splitter.split_text(research_paper_content)
+article_information_chunks = text_splitter.split_text(article_information_content)
+
+print(f"# Chunks in Research Paper: {len(research_paper_chunks)}")
+print(f"# Chunks in Article Document: {len(article_information_chunks)}")
+
+from langchain.embeddings.openai import OpenAIEmbeddings
+import os
+
+os.environ["OPENAI_API_KEY"] = "<YOUR_KEY>"
+embeddings = OpenAIEmbeddings()
+
+from langchain.vectorstores import FAISS
+
+def get_doc_search(text_splitter):
+    
+    return FAISS.from_texts(text_splitter, embeddings)
+
+doc_search_paper = get_doc_search(research_paper_chunks)
+print(doc_search_paper)
+
+
+from langchain.llms import OpenAI
+from langchain.chains.question_answering import load_qa_chain
+chain = load_qa_chain(OpenAI(), chain_type = "map_rerank",  
+                      return_intermediate_steps=True)
+
+def chat_with_file(file_path, query):
+    
+    file_content = extract_file_content(file_path)
+    text_splitter = text_splitter.split_text(file_content)
+    
+    document_search = get_doc_search(text_splitter)
+    documents = document_search.similarity_search(query)
+    
+    results = chain({
+                        "input_documents":documents, 
+                        "question": query
+                    }, 
+                    return_only_outputs=True)
+    answers = results['intermediate_steps'][0]
+    
+    return answers
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

It's good practice to end your files with a newline character to comply with POSIX standards and ensure compatibility across different platforms and tools.
```diff
+ \n
```
Just add a newline at the end of the file.


<!-- This is an auto-generated comment by OSS CodeRabbit -->

Reply:
@coderabbitai can you write test cases for the chat_with_file method?

Reply:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

@Vagif12 Sure, here are some basic test cases for the `chat_with_file` method. These tests use Python's built-in `unittest` module.

```python
import unittest
from unittest.mock import patch, MagicMock

class TestChatWithFile(unittest.TestCase):
    @patch('main_model.extract_file_content')
    @patch('main_model.get_doc_search')
    @patch('main_model.load_qa_chain')
    def test_chat_with_file(self, mock_load_qa_chain, mock_get_doc_search, mock_extract_file_content):
        # Mocking the dependencies
        mock_extract_file_content.return_value = "This is a sample file content"
        mock_get_doc_search.return_value = MagicMock()
        mock_load_qa_chain.return_value = MagicMock()

        # Define a sample query
        query = "What is the content of the file?"

        # Call the function with the sample file path and query
        result = chat_with_file("sample_file_path", query)

        # Check if the dependencies were called correctly
        mock_extract_file_content.assert_called_once_with("sample_file_path")
        mock_get_doc_search.assert_called_once()
        mock_load_qa_chain.assert_called_once()

        # Add more assertions based on your specific requirements and expected results

if __name__ == '__main__':
    unittest.main()
```

Please note that this is a basic test case and it assumes that the `chat_with_file` function is in the global scope. If the function is part of a class, you might need to adjust the test case accordingly. Also, you may want to add more assertions based on your specific requirements and expected results.

<!-- This is an auto-generated reply by OSS CodeRabbit -->